crates/volta-migrate/src/v3.rs

```diff
@@ -0,0 +1,144 @@
+        V3 {
+            home: v3::VoltaHome::new(home),
+        }
+    }
+
+    /// Write the layout file to mark migration to V2 as complete
+    ///
+    /// Should only be called once all other migration steps are finished, so that we don't
+    /// accidentally mark an incomplete migration as completed
+    fn complete_migration(home: v3::VoltaHome) -> Fallible<Self> {
+        debug!("Writing layout marker file");
+        File::create(home.layout_file()).with_context(|| ErrorKind::CreateLayoutFileError {
+            file: home.layout_file().to_owned(),
+        })?;
+
+        Ok(V3 { home })
+    }
+}
+
+impl TryFrom<Empty> for V3 {
+    type Error = VoltaError;
+
+    fn try_from(old: Empty) -> Fallible<Self> {
+        debug!("New Volta installation detected, creating fresh layout");
+
+        let home = v3::VoltaHome::new(old.home);
+        home.create().with_context(|| ErrorKind::CreateDirError {
+            dir: home.root().to_owned(),
+        })?;
+
+        V3::complete_migration(home)
+    }
+}
+
+impl TryFrom<V2> for V3 {
+    type Error = VoltaError;
+
+    fn try_from(old: V2) -> Fallible<Self> {
+        debug!("Migrating from V2 layout");
+
+        let new_home = v3::VoltaHome::new(old.home.root().to_owned());
+        new_home
+            .create()
+            .with_context(|| ErrorKind::CreateDirError {
+                dir: new_home.root().to_owned(),
+            })?;
+
+        // Migrate installed packages to the new workflow
+        migrate_packages(&old.home)?;
+
+        // Complete the migration, writing the V3 layout file
+        let layout = V3::complete_migration(new_home)?;
+
+        // Remove the V2 layout file, since we're now on V3 (do this after writing the V3 file so that we know the migration succeeded)
+        remove_file_if_exists(old.home.layout_file())?;
+
+        Ok(layout)
+    }
+}
+
+fn migrate_packages(old_home: &v2::VoltaHome) -> Fallible<()> {
+    let packages = get_installed_packages(old_home);
+    let mut session = Session::init();
+
+    for package in packages {
+        migrate_single_package(package, &mut session)?;
+    }
+
+    Ok(())
+}
+
+/// Determine a list of all installed packages that are using the legacy package config
+fn get_installed_packages(old_home: &v2::VoltaHome) -> Vec<LegacyPackageConfig> {
+    WalkDir::new(old_home.default_package_dir())
+        .max_depth(2)
+        .into_iter()
+        .filter_map(|res| match res {
+            Ok(entry) => {
+                if entry.file_type().is_file() {
+                    let config = LegacyPackageConfig::from_file(entry.path());
+
+                    if config.is_none() {
+                        debug!("Unable to parse config file: {}", entry.path().display());
+                    }
+
+                    config
+                } else {
+                    None
+                }
+            }
+            Err(error) => {
+                debug!("Error reading directory entry: {}", error);
```

Comment:
Good call! I think having a warning that is shown in the non-verbose case makes sense, where we can include a CTA to run `volta install <package>` manually if it's needed.

Reply:
Actually looking at it, this error is an error even reading a directory entry. We can't provide a meaningful warning message for that, because if we can't read the directory entry, we don't even know what file was (supposedly) there to determine the package name. That said, above when we parse the package config, we _can_ show a warning if that fails.

Reply:
Added a warning if we're unable to parse a config file, asking the user to run `volta install <package>` to manually migrate.